### ISSUE TITLE
[java] Better Lombok support

### ISSUE THREAD

**Issue by oowekyala** [pmd/pmd#3119]
Many rules have some form of "lombok awareness", in that they ignore some constructs if they have lombok annotations. But PMD does not really handle the desugaring that lombok performs in a way, that all rules can benefit transparently.

The main issues are:
* `lombok.val` should be treated like the `var` keyword of java 10. #3118
* Type resolution in PMD 7 uses the AST as a source of symbols (methods, etc) to perform eg overload resolution. Those symbols are implicit when you use lombok. (eg `@Getter` declares a getter method), and for now are not picked up. This means, that type resolution will be much more patchy on lombok codebases, in the current state. A nice solution, would be to generate some of the symbols that correspond to annotations (the same way [we add a generated symbol for eg `Enum.valueOf`](https://github.com/pmd/pmd/blob/5ca13f1bdd7338a2ea7121eb4ba2931fccad3c51/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/symbols/internal/ast/AstClassSym.java#L121-L124) in enum declarations). I think fully replicating the lombok desugaring is too complicated and is a non-goal. We should only generate symbols that are usable from the source language.

  For instance, the [`@Synchronized`](https://objectcomputing.com/resources/publications/sett/january-2010-reducing-boilerplate-code-with-project-lombok#synchronized) annotation generates a private lock field. Generating a symbol for this field is a non-goal, because it's an internal detail. It cannot mess up type resolution as it's not used in the source we're processing, only in the modified AST that lombok generates. Same thing for all annotations that just alter the AST locally, like `@Cleanup`.
  
  OTOH, `@Getter` is something we could care about, because some dependent code may be using the getter method. If we're not aware of this, then that dependent code will be mistyped.

I think that second item may only be implemented in PMD 7. The first item could also be implemented in pmd 6 by a motivated person.

**Comment by adangel** [comment]
> Those symbols are implicit when you use lombok. (eg `@Getter` declares a getter method), and for now are not picked up. This means, that type resolution will be much more patchy on lombok codebases, in the current state. A nice solution, would be to generate some of the symbols that correspond to annotations 

Not sure, if this is needed. When lombok is used, the produced bytecode contains the generated getter method. So we should basically pick it up via ASM, when the auxclasspath is provided. At least, it should work, when a dependent file is analyzed. If we analyze the file, which has `@Getter`, we have the situation, that the bytecode has more methods than we see in the source...

**Comment by oowekyala** [comment]
> At least, it should work, when a dependent file is analyzed.

Yes, but the symbol implementation for those symbols that are declared in the current file exclusively rely on the AST and not the class file. Maybe we could use classpath info, but it would be hard to determine which symbols are equal and merge them, or handle inconsistencies and out-of-sync source

> we have the situation, that the bytecode has more methods than we see in the source

I don't think this is a problem, as we already add synthetic method symbols for eg enums and record components

**Comment by Frederick888** [comment]
Hi, new to this project, just wondering if supporting lombok in `InvalidLogMessageFormat` falls into the scope of this issue?

For example, pmd reports 'missing arguments' on

```java
public class Book {
    private static final Logger LOGGER = LoggerFactory.getLogger(Book.class);

    public void logMe() {
        LOGGER.info("Hello, %s!");
    }
}
```
...but not on
```java
@Slf4j
public class Book {
    public void logMe() {
        log.info("Hello, %s!");
    }
}
```
Thanks :)

**Comment by mrclmh** [comment]
Could it be an option to [delombok](https://projectlombok.org/features/delombok) the sources as a preprocessing step?

**Comment by jsotuyod** [comment]
unlikely…

* it would require to ship lombok with PMD (potential version conflicts)
* it would change the user's codebase (big nono)
* it would change the analyzed source, leading to mismatches in reported code positions…
